gnat2goto/driver/tree_walk.adb

```diff
@@ -3531,36 +3539,26 @@ package body Tree_Walk is
    end Do_Op_Minus;
 
    -------------------------
-   --      Do_Op_Or       --
-   -------------------------
-
-   function Do_Op_Or (N : Node_Id) return Irep is
-      LHS_Bool_Value : constant Irep := Do_Expression (Left_Opnd (N));
-      RHS_Bool_Value : constant Irep := Do_Expression (Right_Opnd (N));
-      Cast_LHS_To_Integer : constant Irep :=
-        Make_Op_Typecast (Op0 => LHS_Bool_Value,
-                          Source_Location => Sloc (N),
-                          I_Type => Make_Signedbv_Type (Make_Nil_Type, 32));
-      Cast_RHS_To_Integer : constant Irep :=
-        Make_Op_Typecast (Op0 => RHS_Bool_Value,
-                          Source_Location => Sloc (N),
-                          I_Type => Make_Signedbv_Type (Make_Nil_Type, 32));
-      R : constant Irep := Make_Op_Bitor (Lhs => Cast_LHS_To_Integer,
-                                          Rhs => Cast_RHS_To_Integer,
-                                          Source_Location => Sloc (N),
-                                          I_Type =>
-                                            Get_Type (Cast_LHS_To_Integer));
-   begin
-      return Make_Op_Typecast (Op0 => R,
-                               Source_Location => Sloc (N),
-                               I_Type => Make_Bool_Type);
-   end Do_Op_Or;
-
-   -------------------------
-   --      Do_Op_And      --
+   --      Do_Bit_Op      --
    -------------------------
 
-   function Do_Op_And (N : Node_Id) return Irep is
+   --  We identified that the constructor for operator `or` and operator
+   --  `and` where pretty much the same, with the only difference being
+   --  the constructor being called. So to avoid needless duplication,
+   --  we simplified it to a single function that does the same thing,
+   --  and just calls the appropriate constructor via a function pointer.
+   --
+   --  This produces the following code in pseudocode
+   --    A or B  (assume A => True and B => True)
+   --  --
+   --    int intA = (int) A;       // 1
+   --    int intB = (int) B;       // 1
+   --    int intC = A | B          // bitwise `or` or `and`
```

Comment:
I'd just say `bitop` instead of `|` followed by `// bitop can be bitor, bitand`

and theoretically also `xor` and things like that.

Reply:
Good point.

Reply:
Also, the actual generated code looks more like `(bool) ((int) A bitop (int) B)`?

Reply:
Yes, indeed, I wanted to make it as simple as possible, and try to simulate the sequence of the calls happening inside the function.